xla/pjrt/host_memory_spaces.h

```diff
@@ -91,6 +91,72 @@ class PinnedHostMemorySpace : public PjRtMemorySpace {
   std::string to_string_;
 };
 
+class DeviceMemorySpace : public PjRtMemorySpace {
+ public:
+  static constexpr absl::string_view kKind = "device";
+  static const int kKindId;
+
+  DeviceMemorySpace(int id, PjRtDevice* device);
+
+  PjRtClient* client() const override { return client_; }
+
+  absl::Span<PjRtDevice* const> devices() const override {
+    return absl::Span<PjRtDevice* const>(&device_, device_ != nullptr ? 1 : 0);
+  }
+
+  int id() const override { return id_; }
+
+  absl::string_view kind() const override { return kKind; }
+
+  int kind_id() const override { return kKindId; }
+
+  absl::string_view DebugString() const override { return debug_string_; }
+
+  absl::string_view ToString() const override { return to_string_; }
+
+  void AttachDevice(PjRtDevice* device) { device_ = device; }
+
+ private:
+  int id_;
+  PjRtClient* client_ = nullptr;
+  PjRtDevice* device_ = nullptr;
+  std::string debug_string_;
+  std::string to_string_;
+};
+
+class NcclMemorySpace : public PjRtMemorySpace {
```

Comment:
This module and this folder only holds memory space representations which can be shared across different PJRT backends.

GPU-specific memory space should be placed in a GPU folder. I guess other than NV GPU, nobody will use the NcclMemorySpace anymore.

BTW, the name is pretty confusing. I can imagine the memory space name can be closer to the hardware name. NCCL is a library name and from a reader perspective, I don't know what that is referring to.